src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.cs

```diff
@@ -110,44 +121,126 @@ protected override void Dispose(bool disposing)
+            IntPtr headerPtr = IntPtr.Zero;
+            bool referenceTaken = false;
+
+#if DEBUG
+            if (Logging.On) Logging.Enter(Logging.Http, this, "StartRequest", state.RequestMessage);
+#endif
 
             if (state.CancellationToken.IsCancellationRequested)
             {
-                state.TrySetCanceled();
+                state.TrySetCanceled(state.CancellationToken);
                 return;
             }
 
-            var cancellationTokenRegistration = state.CancellationToken.Register(x => ((RequestCompletionSource) x).TrySetCanceled(), state);
+            var cancellationTokenRegistration = state.CancellationToken.Register(
+                s =>
+                {
+                    var rcs = (RequestCompletionSource)s;
+                    rcs.TrySetCanceled(rcs.CancellationToken);
+                },
+                state);
 
             try
             {
-                requestHandle = CreateRequestHandle(state);
+                // Prepare context objects
+                requestStateHandle = GCHandle.Alloc(state);
+                responseMessage = new CurlResponseMessage(state.RequestMessage);
+                state.ResponseMessage = responseMessage;
+                requestHandle = CreateRequestHandle(state, requestStateHandle, out headerPtr);
 
-                state.CancellationToken.ThrowIfCancellationRequested();
-                int result = Interop.libcurl.curl_easy_perform(requestHandle);
-                if (CURLcode.CURLE_OK != result)
+                if (state.CancellationToken.IsCancellationRequested)
                 {
-                    throw new HttpRequestException(SR.net_http_client_execution_error, GetCurlException(result));
+                    state.TrySetCanceled(state.CancellationToken);
+                    return;
                 }
 
-                // TODO: Handle requests with body
+                if (state.RequestMessage.Content != null)
+                {
+#if DEBUG
+                    Logging.PrintInfo(Logging.Http, this, "Request content for " + Logging.GetObjectLogHash(responseMessage.ContentStream) + " : " + state.RequestMessage.Content.ReadAsStringAsync().Result);
+#endif
+
+                    Stream requestContentStream =
+                        await state.RequestMessage.Content.ReadAsStreamAsync().ConfigureAwait(false);
+                    if (state.CancellationToken.IsCancellationRequested)
+                    {
+                        state.TrySetCanceled(state.CancellationToken);
+                        return;
+                    }
+                    state.RequestContentStream = requestContentStream;
+                }
 
-                state.CancellationToken.ThrowIfCancellationRequested();
-                responseMessage = CreateResponseMessage(requestHandle, state.RequestMessage);
+                requestHandle.DangerousAddRef(ref referenceTaken);
+                var performTask = Task.Run(() =>
```

Comment:
The main issue I am getting blocked on is identifying when all headers have been received. This can be signaled in 2 ways:
1. In case of a HEAD request, curl_*_perform returns completion. (In sync variant, function call returns and in async variant, an out parameter needs to be checked)
2. The response body callback is received. We are still investigating if this is also reliable.

So a solution based only on callbacks doesn't seem likely